src/input/isf.c

```diff
@@ -0,0 +1,796 @@
+#define MAX_WAVEFORM_STRING_SIZE 10
+
+/* Maximum number of bytes per sample. */
+#define MAX_INT_BYTNR 8
+#define FLOAT_BYTNR 4
+
+/* Size of buffer in which byte order and data format strings are stored. */
+#define BYTE_ORDER_BUFFER_SIZE 4
+#define DATA_FORMAT_BUFFER_SIZE 3
+
+/* Byte order */
+enum byteorder {
+	LSB = 0,
+	MSB = 1,
+};
+
+/* Format, i.e. RI (signed integer), RP (unsigned integer) or FP (floating point) */
+enum format {
+	RI = 0,
+	RP = 1,
+	FP = 2,
+};
+
+/* Waveform type, i.e. analog or radar frequency (RF) */
+enum waveform_type {
+	ANALOG = 0,
+	RF_FD = 1,
+};
+
+union floating_point {
+	float f;
+	uint32_t i;
+};
+
+struct context {
+	gboolean started;
+	gboolean create_channel;
+	gboolean found_data_section;
+	float yoff;
+	float yzero;
+	float ymult;
+	float xincr;
+	unsigned int bytnr;
+	enum byteorder byte_order;
+	enum format bn_fmt;
+	enum waveform_type wfmtype;
+	char channel_name[MAX_CHANNEL_NAME_SIZE];
+};
+
+/*
+ * Header items used to process the input file.
+ *
+ * Parameter WFID is optional, the rest are required.
+ */
+enum header_items_enum {
+	YOFF = 0,
+	YZERO = 1,
+	YMULT = 2,
+	XINCR = 3,
+	BYTNR = 4,
+	BYTE_ORDER = 5,
+	BN_FMT = 6,
+	WFID = 7,
+	WFMTYPE = 8,
+	ENCODING = 9,
+};
+
+/* Strings searched for in the file header representing the header items. */
+static const char *header_items[] = {
+	[YOFF] = "YOFF ",
+	[YZERO] = "YZERO ",
+	[YMULT] = "YMULT ",
+	[XINCR] = "XINCR ",
+	[BYTNR] = "BYT_NR ",
+	[BYTE_ORDER] = "BYT_OR ",
+	[BN_FMT] = "BN_FMT ",
+	[WFID] = "WFID ",
+	[WFMTYPE] = "WFMTYPE ",
+	[ENCODING] = "ENCDG ",
+};
+
+/* Find the header item string in the header. */
+static char *find_item(const char *buf, size_t buflen, const char *item)
+{
+	return g_strstr_len(buf, buflen, item);
+}
+
+/* Find curve which marks the end of the header and the start of the data. */
+static char *find_data_section(GString *buf)
+{
+	const char curve[] = "CURVE #";
+
+	char *data_ptr;
+	size_t offset, metadata_length;
+
+	data_ptr = g_strstr_len(buf->str, buf->len, curve);
+	if (data_ptr == NULL)
+		return NULL;
+
+	data_ptr += strlen(curve);
+	offset = data_ptr - buf->str;
+	if (offset >= buf->len)
+		return NULL;
+
+	/* Curve metadata length is an ASCII byte, hence -48. */
+	metadata_length = (size_t) *data_ptr - 48;
+	data_ptr += 1 + metadata_length;
+	offset = (size_t) (data_ptr - buf->str);
+
+	if (offset >= buf->len)
+		return NULL;
+
+	return data_ptr;
+}
+
+/* Check if the entire header is loaded and can be processed. */
+static gboolean has_header(GString *buf)
+{
+	return find_data_section(buf) != NULL;
+}
+
+/* Locate and extract the channel name in the header. */
+static void extract_channel_name(struct context *inc, const char *buf, size_t buflen)
+{
+	size_t i, channel_ix;
+
+	channel_ix = 0;
+	/* ISF WFID looks something like WFID "Ch1, ..."; hence we must skip character '"' */
+	i = 1;
+	while (i < buflen && buf[i] != ',' && buf[i] != '"' && channel_ix < MAX_CHANNEL_NAME_SIZE - 1)
```

Comment:
not a fan of this : very long line, with many conditions, and the following instruction block isn't braced. Makes this hard to process if editor wraps the line due to limited width

Reply:
I braced the instruction block and wrapped the condition line.